File: src/entities/Animation.java
package entities;

import org.newdawn.slick.Image;
import org.newdawn.slick.SpriteSheet;

public class Animation {
	private Image[] frames;
	
	private int[] frameOrder;
	
	private int currentFrame;
	
	public Animation(SpriteSheet sheet) {
		//hush bb is ok
		//this assumes none are empty sorry
		Image[] topsecret = new Image[sheet.getVerticalCount() * sheet.getHorizontalCount()];
		int tileWidth = sheet.getWidth() / sheet.getHorizontalCount();
		int tileHeight = sheet.getHeight() / sheet.getVerticalCount();
		for (int j = 0; j < sheet.getVerticalCount(); j++) {
			for (int i = 0; i < sheet.getHorizontalCount(); i++) {
				int n = j * sheet.getHorizontalCount() + i;
				int x = i * tileWidth, y = j * tileHeight;
				int ox = x + tileWidth, oy = y + tileHeight;
				topsecret[n] = sheet.getSprite(i, 0);
			}
		}
		
		this.frames = topsecret;
		this.frameOrder = new int[frames.length];
		for (int i = 0; i < frames.length; i++) {
			frameOrder[i] = i;
		}
	}
	
	public Animation(int[] frameOrder, Image... frames) {
		this.frameOrder = frameOrder;
		this.frames = frames;
	}
	
	public Animation(Image... frames) {
		this.frames = frames;
		this.frameOrder = new int[frames.length];
		for (int i = 0; i < frames.length; i++) {
			frameOrder[i] = i;
		}
	}
	
	public void nextFrame() {
		currentFrame++;
		if (currentFrame >= frameOrder.length) {
			currentFrame = 0;
		}
	}
	
	public void setFrame(int frameNum) {
		currentFrame = frameNum;
	}
	
	public Image getCurrentFrame() {
		return frames[frameOrder[currentFrame]];
	}
	
	public int getCurrentFrameNum() {
		return currentFrame;
	}
}

File: src/entities/Animation.kt
package entities

import org.newdawn.slick.Image

class Animation {
    private var frames: ArrayList<Image> = ArrayList()

    private var frameOrder: IntArray? = null

    private var currentFrameNum: Int = 0

    val currentFrame: Image
        get() = frames[frameOrder!![currentFrameNum]]

    constructor(frameOrder: IntArray, vararg frames: Image) {
        this.frameOrder = frameOrder
        this.frames = ArrayList(frames.toMutableList())
    }

    constructor(vararg frames: Image) {
        //TODO("Fix this awful black magic fuckery right here.")
        this.frames = ArrayList(frames.toMutableList())
        this.frameOrder = IntArray(frames.size)
        for (i in frames.indices) {
            frameOrder!![i] = i
        }
    }

    fun nextFrame() {
        currentFrameNum++
        if (currentFrameNum >= frameOrder!!.size) {
            currentFrameNum = 0
        }
    }

    fun setFrame(frameNum: Int) {
        currentFrameNum = frameNum
    }
}
